CRM/Sqltasks/BAO/SqlTask.php

```diff
@@ -0,0 +1,717 @@
+   *
+   * @return void
+   */
+  public function unarchive() {
+    $this->updateAttributes([ 'archive_date' => NULL ]);
+  }
+
+  /**
+   * Update task data
+   *
+   * @param array $params
+   * @param bool $save
+   * @return void
+   */
+  public function updateAttributes($params, $save = TRUE) {
+    foreach ($params as $key => $value) {
+      if (in_array($value, [NULL, ''], TRUE) && !self::fields()[$key]['required']) {
+        $this->$key = NULL;
+        continue;
+      }
+
+      switch ($key) {
+        case 'abort_on_error':
+        case 'enabled':
+        case 'input_required': {
+          $value = $value === '' ? FALSE : $value;
+
+          if (!in_array($value, [TRUE, FALSE, 1, 0, '1', '0'], TRUE)) {
+            throw new Exception("Attribute '$key' must be a boolean");
+          }
+
+          $value = (bool) $value;
+
+          if (
+            $key === 'enabled'
+            && !$value
+            && isset($this->id)
+            && !empty(self::getDependentTasks($this->id, TRUE))
+          ) {
+            throw new Exception("Task can not be disabled, other tasks depend on it");
+          }
+
+          $this->$key = $value;
+          break;
+        }
+
+        case 'parallel_exec': {
+          if (!in_array($value, [0, 1, 2, '0', '1', '2'], TRUE)) {
+            throw new Exception("Attribute '$key' must be one of 0, 1 or 2");
+          }
+
+          $this->$key = (int) $value;
+          break;
+        }
+
+        case 'last_runtime':
+        case 'weight': {
+          if (!self::isPositiveInt($value)) {
+            throw new Exception("Attribute '$key' must be a positive integer");
+          }
+
+          $this->$key = (int) $value;
+          break;
+        }
+
+        case 'category':
+        case 'description':
+        case 'name':
+        case 'run_permissions': {
+          $this->$key = (string) $value;
+          break;
+        }
+
+        case 'archive_date':
+        case 'last_execution':
+        case 'running_since': {
+          if (strtotime($value) === FALSE) {
+            throw new Exception("Attribute '$key' must be a date");
+          }
+
+          $this->$key = date('Y-m-d H:i:s', strtotime($value));
+          break;
+        }
+
+        case 'config': {
+          $value = is_string($value) ? json_decode($value, TRUE) : $value;
+          $config = self::validateConfiguration($value);
+          $this->$key = json_encode($config);
+          break;
+        }
+
+        case 'scheduled': {
+          if (!in_array($value, self::$schedulingOptions)) {
+            throw new Exception("Attribute '$key' must be a valid interval");
+          }
+
+          $this->$key = (string) $value;
+          break;
+        }
+      }
+    }
+
+    if ($save) {
+      // Temporarily change DB_DataObject configuration to enable saving of NULL values
+      global $_DB_DATAOBJECT;
+      $disable_null_strings = $_DB_DATAOBJECT['CONFIG']['disable_null_strings'] ?? NULL;
```

Comment:
I remember discussing this before, but don't remember the conclusion. 😅

Based on other code I found in Core BAOs, it looks like the canonical way of representing "please write a null value to this field in the database" is to use `'null'` (as a string). It doesn't look like any code in Core reconfigures `disable_null_strings`, so I'd lean towards doing `$this->$key = 'null';` in L532 instead, unless that doesn't work for some reason?

(I'm a bit worried about this approach because our call to `$this->save();` may trigger Pre/Post hooks registered by other extensions, which may break due to the side-effect we added by changing the `disable_null_strings` configuration. It's not likely that anyone would want to hook into the SqlTask BAO, but if the solution is as simple as using a (admittedly, very ugly) `null` string, I'd rather do that.)